File: packages/db/prisma/seed/capabilities.ts
/**
 * Data Partner capability metadata — v6 §5.1, §7.
 *
 * §5.1 is what a Partner publishes to Oolix: which standardized attributes they
 * can answer questions about, over which geographies and channels. §17 draws
 * the line hard — Oolix stores "Partner capability metadata, not Partner
 * customer records or Partner local field names". So nothing here names a
 * column, a table or a person. `payment_method` is a taxonomy key; the fact
 * that Travel A stores it as `pay_mode` lives in the Agent's own config file
 * and never reaches this database.
 *
 * The two Partners deliberately publish DIFFERENT coverage. That is what makes
 * §7's asymmetry visible in the demo rather than theoretical: a travel audience
 * with a REQUIRED `recent_booking` rule is COMPATIBLE for Travel A and
 * INCOMPATIBLE for Rewards B — not "lower scoring", but unanswerable.
 */
import type { PrismaClient } from '../../src/generated/prisma/client.js';

/** §4: the operators each data type supports, mirrored from the taxonomy. */
const NUMBER_OPS = ['EQ', 'IN', 'LTE', 'GTE', 'BETWEEN'];
const ENUM_OPS = ['EQ', 'IN'];
const BOOLEAN_OPS = ['EQ'];

interface CapabilitySeed {
  partnerOrgId: string;
  label: string;
  /** Taxonomy keys only. Never a local field name (§17). */
  attributes: Array<{ attribute_key: string; operators: string[] }>;
  geographies: string[];
  channels: string[];
  /**
   * §11: the Agent reports which local mapping version it compiled against, so
   * a reach estimate can be traced to the mapping that produced it. It must
   * match the `mapping_version` in that Partner's agent config.
   */
  mappingVersion: number;
}

const attr = (keys: string[], operators: string[]) =>
  keys.map((attribute_key) => ({ attribute_key, operators }));

export const PARTNER_CAPABILITIES: CapabilitySeed[] = [
  {
    // Travel A runs the local Agent in this environment, and its attribute
    // view covers the full Appendix A catalogue — so its published capability
    // matches partner-agent/config.example.yaml exactly. Publishing more than
    // the Agent can map would make Oolix send rules the Agent must refuse.
    partnerOrgId: '22222222-2222-4222-8222-222222222222',
    label: 'Travel A',
    attributes: [
      ...attr(
        [
          'age',
          'purchase_recency_days',
          'purchase_frequency',
          'booking_recency_days',
          'active_user_days',
        ],
        NUMBER_OPS,
      ),
      ...attr(
        [
          'gender',
          'country',
          'state_region',
          'city',
          'purchase_category',
          'payment_method',
          'domestic_international',
          'loyalty_tier',
        ],
        ENUM_OPS,
      ),
      ...attr(['online_shopper', 'recent_booking', 'app_active'], BOOLEAN_OPS),
    ],
    geographies: ['IN'],
    // §7 / §84: Meta and Google are absent rather than declared-and-disabled.
    // A capability a Buyer can see is one they can select, and §32 says not to
    // show an external channel as selectable until eligibility is proven.
    channels: ['PARTNER_WEB', 'PARTNER_APP'],
    mappingVersion: 8,
  },
  {
    // Rewards B is a retail rewards programme: rich commerce data, no travel
    // history at all. Its Agent has no `has_recent_trip` column to map, so
    // declaring the attribute would be a lie the Agent could not honour.
    partnerOrgId: '33333333-3333-4333-8333-333333333333',
    label: 'Rewards B',
    attributes: [
      ...attr(['age', 'purchase_recency_days', 'active_user_days'], NUMBER_OPS),
      ...attr(
        ['gender', 'country', 'state_region', 'purchase_category', 'payment_method'],
        ENUM_OPS,
      ),
      ...attr(['online_shopper', 'app_active'], BOOLEAN_OPS),
    ],
    geographies: ['IN'],
    channels: ['PARTNER_WEB'],
    mappingVersion: 3,
  },
];

/**
 * Publish version 1 of each Partner's capabilities.
 *
 * Idempotent by (partner, capability_version) so re-seeding never creates a
 * second version. A real re-publish goes through
 * `POST /v1/partner/capabilities`, which mints a NEW version — §16 keeps old
 * versions readable so a past approval can still be explained.
 */
export async function seedPartnerCapabilities(prisma: PrismaClient): Promise<void> {
  for (const cap of PARTNER_CAPABILITIES) {
    await prisma.partnerCapability.upsert({
      where: {
        partnerOrgId_capabilityVersion: {
          partnerOrgId: cap.partnerOrgId,
          capabilityVersion: 1,
        },
      },
      create: {
        partnerOrgId: cap.partnerOrgId,
        capabilityVersion: 1,
        attributesJson: cap.attributes.map((a) => ({ ...a, status: 'AVAILABLE' })),
        geographiesJson: cap.geographies,
        channelsJson: cap.channels,
        status: 'ACTIVE',
        mappingVersion: cap.mappingVersion,
      },
      update: {},
    });
  }

  const summary = PARTNER_CAPABILITIES.map(
    (c) => `${c.label}: ${c.attributes.length} attributes`,
  ).join(', ');
  console.log(`[seed] partner capabilities (${summary})`);
}

File: partner/agent/internal/managed/capabilities.go
package managed

import (
	"errors"

	"github.com/oolix/partner-agent/internal/standard"
)

// Capability is one attribute the Agent offers to answer questions about.
type Capability struct {
	AttributeKey string   `json:"attribute_key"`
	Operators    []string `json:"operators"`
	Status       string   `json:"status"`
}

// Capabilities is everything the Agent tells Oolix. It names standard
// attributes and nothing else: never a column, a value or a customer.
type Capabilities struct {
	Attributes     []Capability `json:"attributes"`
	Geographies    []string     `json:"geographies"`
	Channels       []string     `json:"channels"`
	MappingVersion int          `json:"mapping_version"`
}

// ErrNothingToPublish means no attribute the Partner chose had any readable
// value in the copy.
var ErrNothingToPublish = errors.New(
	"nothing to publish: none of the chosen attributes had a readable value in the last sync")

// BuildCapabilities lists the attributes the Partner chose to publish that the
// last good sync actually filled. An attribute whose column turned out empty
// is left out: a Buyer targeting it would reach nobody.
func BuildCapabilities(m Mapping, last *Report) (Capabilities, error) {
	caps := Capabilities{
		Geographies:    []string{"IN"},
		Channels:       m.Channels,
		MappingVersion: MappingVersion,
	}
	if len(caps.Channels) == 0 {
		caps.Channels = []string{"PARTNER_WEB"}
	}
	if last == nil {
		return caps, ErrNothingToPublish
	}
	provides := ActivityProvides(m)
	for _, a := range standard.Attributes {
		if provides[a.Key] {
			if withheld(m, a.Key) || last.Attributes[a.Key].Filled == 0 {
				continue
			}
		} else if am, ok := m.Attributes[a.Key]; !ok || am.Column == "" || !am.Publish ||
			last.Attributes[a.Key].Filled == 0 {
			continue
		}
		caps.Attributes = append(caps.Attributes, Capability{
			AttributeKey: a.Key, Operators: a.Operators, Status: "AVAILABLE",
		})
	}
	if len(caps.Attributes) == 0 {
		return caps, ErrNothingToPublish
	}
	return caps, nil
}

// withheld says whether the Partner chose not to offer an attribute their
// orders or bookings table supplies.
func withheld(m Mapping, key string) bool {
	for _, am := range []*ActivityMapping{m.Orders, m.Bookings} {
		if am != nil && contains(am.Withhold, key) {
			return true
		}
	}
	return false
}

// Withdrawn is the same list marked unavailable: what is published when the
// Partner deletes the copy, so no Buyer is offered an audience that cannot be
// served.
func (c Capabilities) Withdrawn() Capabilities {
	out := c
	out.Attributes = make([]Capability, len(c.Attributes))
	for i, a := range c.Attributes {
		a.Status = "UNAVAILABLE"
		out.Attributes[i] = a
	}
	return out
}

// Keys lists the attribute keys, for the status page.
func (c Capabilities) Keys() []string {
	out := make([]string, len(c.Attributes))
	for i, a := range c.Attributes {
		out[i] = a.AttributeKey
	}
	return out
}
